# Foldmason
Foldmason builds multiple alignments of large structure sets.

## Publications
[van Kempen M, Kim S, Tumescheit C, Mirdita M, Lee J, Gilchrist C, SÃ¶ding J, and Steinegger M. Fast and accurate protein structure search with Foldseek. Nature Biotechnology, doi:10.1038/s41587-023-01773-0 (2023)](https://www.nature.com/articles/s41587-023-01773-0)

[![build workflow](https://github.com/steineggerlab/foldmason/actions/workflows/build.yml/badge.svg)](https://github.com/steineggerlab/foldmason/actions/workflows/build.yml)

# Table of Contents

- [Foldmason](#foldmason)
- [Webserver](#webserver)
- [Installation](#installation)
- [Documentation](#documentation)
- [Quick Start](#quick-start)
  - [Multiple alignment](#multiple-alignment)
    - [Output](#output)
    - [Important Parameters](#important-parameters)
  - [Databases](#databases)
    - [Create Custom Databases and Indexes](#create-custom-databases-and-indexes)
- [Main Modules](#main-modules)
- [Examples](#examples)

## Webserver 
Align your protein structures quickly using our Foldmason webserver: ...
<!-- Search your protein structures against the [AlphaFoldDB](https://alphafold.ebi.ac.uk/) and [PDB](https://www.rcsb.org/) in seconds using our Foldseek webserver: [search.foldseek.com](https://search.foldseek.com) ðŸš€ -->

## Installation
```
# Linux AVX2 build (check using: cat /proc/cpuinfo | grep avx2)
wget https://mmseqs.com/foldmason/foldmason-linux-avx2.tar.gz; tar xvzf foldmason-linux-avx2.tar.gz; export PATH=$(pwd)/foldmason/bin/:$PATH

# Linux SSE2 build (check using: cat /proc/cpuinfo | grep sse2)
wget https://mmseqs.com/foldmason/foldmason-linux-sse2.tar.gz; tar xvzf foldmason-linux-sse2.tar.gz; export PATH=$(pwd)/foldmason/bin/:$PATH

# Linux ARM64 build
wget https://mmseqs.com/foldmason/foldmason-linux-arm64.tar.gz; tar xvzf foldmason-linux-arm64.tar.gz; export PATH=$(pwd)/foldmason/bin/:$PATH

# MacOS
wget https://mmseqs.com/foldmason/foldmason-osx-universal.tar.gz; tar xvzf foldmason-osx-universal.tar.gz; export PATH=$(pwd)/foldmason/bin/:$PATH

# Conda installer (Linux and macOS)
conda install -c conda-forge -c bioconda foldmason
```
Other precompiled binaries for ARM64 amd SSE2 are available at [https://mmseqs.com/foldmason](https://mmseqs.com/foldmason).

<!-- ## Documentation
Many of Foldseek's modules (subprograms) rely on MMseqs2. For more information about these modules, refer to the [MMseqs2 wiki](https://github.com/soedinglab/MMseqs2/wiki). For documentation specific to Foldseek, checkout the Foldseek wiki [here](https://github.com/steineggerlab/foldseek/wiki).
 -->

## Quick start

### Multiple alignment
The `easy-msa` module allows you to align multiple query structures formatted in PDB/mmCIF format (flat or gzipped). By default it outputs the alignment as a [FASTA-format file](#fasta-alignment) as well as an interactive [HTML](#interactive-html) output.

    foldmason easy-msa <PDB/mmCIF files> result.fasta tmpFolder
    
#### Output
##### FASTA alignment
Foldmason generates alignments in FASTA-format.
Alignments using the 3Di alphabet can be generated by specifying the output mode `--output-mode 1`.

##### Interactive HTML
Foldmason can generate a HTML MSA visualisation by specifying `--lddt-html` in the `msa2lddt` module.

```
foldmason msa2lddt myDb result.fasta --lddt-html result.html
```

This is generated automatically when using the `easy-msa`. The following will produce `result.fasta` and `result.html`.

```
foldmason easy-msa <PDB/mmCIF files> result.fasta tmpFolder
```

<p align="center"><img src="./.github/html.gif" height="400"/></p>

#### Important parameters

| Option             | Category  | Description                                                                                               |
|--------------------|-----------|-----------------------------------------------------------------------------------------------------------|
| `--gap-open`       | Alignment | Gap opening penalty (default: 10)
| `--gap-extend`     | Alignment | Gap extension penalty (default: 1)
| `--refine-iters`   | Alignment | Number of refinement iterations to run after initial alignment (default: 0)
| `--output-mode`    | Alignment | 0: Amino acids, 1: 3Di alphabet (default: 0)
| `--pair-threshold` | Scoring   | Maximum proportion of gaps in column threshold for LDDT calculation (default: 0.0)

#### Create custom databases and indexes
The structure database can be pre-processed by `createdb`. This make sense if searched multiple times. 
 
```
foldmason createdb example/ structureDB
```

## Main Modules
- `easy-msa`          multiple alignment workflow from structure files
- `structuremsa`      multiple alignment from structure database
- `msa2lddt`          calculate structure-based score (LDDT) of a MSA 
- `refinemsa`         iterative MSA refinement

## Examples
### Basic MSA workflow
The easiest way to use Foldmason is to use the `easy-msa` workflow like so:

```
foldmason easy-msa <PDB/mmCIF files> result.fasta tmpFolder
```

By default, `easy-msa` produces the multiple alignment in FASTA format (`result.fasta`), as well as the interactive HTML visualisation (`result.html`).
This is the equivalent of the following sequence of commands:

```
foldmason createdb <PDB/mmCIF files> myDb
foldmason structuremsa myDb result.fasta
foldmason msa2lddt myDb result.fasta --lddt-html result.html
```

### Aligning large data sets
Foldmason can use the clustering capabilities of Foldseek to pre-cluster input structures before alignment by specifying `--precluster`,
allowing for alignments of large sets of proteins.

```
foldmason easy-msa <PDB/mmCIF files> result.fasta tmpFolder --precluster
```

### Computing LDDT of an externally created MSA
The `msa2lddt` module computes an average [Local Distance Difference Test (LDDT) score](https://doi.org/10.1093/bioinformatics/btt473)
over the length of an MSA. This is done automatically in the `easy-msa` workflow, but `msa2lddt` can be called separately to compute
the LDDT of any given alignment, so long as the structures in the MSA are present in the given structure database.

```
foldmason msa2lddt myDb result.fasta --lddt-html result.html
```

Average MSA LDDT is calculated by computing per-column LDDT scores of every pair of sequences in the MSA, and then averaging them over the length of the MSA.
Sequence comparison order is determined by database keys, thus scores for MSAs produced by different tools are comparable if specifying the same structure database
in `msa2lddt`.

### MSA Refinement
The `refinemsa` module refines an MSA by iteratively splitting and re-aligning it, saving a resulting MSA when an increase in average LDDT is detected.

```
foldmason refinemsa myDb result.fasta refined.fasta --refine-iters 1000
```

Refinement can be run automatically in the `easy-msa` workflow by specifying the `--refine-iters` argument.
